include/ofi_hmem.h

```diff
@@ -105,17 +101,17 @@ hsa_status_t ofi_hsa_amd_reg_dealloc_cb(void *ptr,
 					hsa_amd_deallocation_callback_t cb,
 					void *user_data);
 
-hsa_status_t ofi_hsa_amd_memory_lock(void *host_ptr, size_t size,
-				     hsa_agent_t *agents, int num_agents,
-				     void **agent_ptr);
-hsa_status_t ofi_hsa_amd_memory_unlock(void *host_ptr);
-
 #endif /* HAVE_ROCR */
 
 struct ofi_hmem_ops {
 	bool initialized;
 	int (*init)(void);
 	int (*cleanup)(void);
+	int (*async_copy_to_hmem)(uint64_t device, void *dest, const void *src,
+			    size_t size, void **stream);
+	int (*async_copy_from_hmem)(uint64_t device, void *dest,
+			const void *src, size_t size, void **stream);
+	int (*async_copy_query)(void *stream);
```

Comment:
My first concern here is the name 'stream', which is confusing when considering the more generic hmem APIs.  The way I envision an async operation working is for the *caller* to pass in some sort of 'event'.  It can then wait or query on whether that event has completed.

Having the function return an event opens up issues with managing the lifetime of the object.  It's clearer if the caller allocates, uses, and frees an event, rather than having the allocation and free hidden without the user of the object knowing when it is freed.

Reply:
@shefty, I'm trying to follow the change you'd like to see here. Currently the event is specific to ROCR, because the signals are ROCR specific. Are you talking about creating an event abstraction, which is populated by the the ROCR specific code, but allocated by the caller?

Reply:
It looks like cuda uses a similar name, sigh...
```
cudaMemcpyAsync(dst, src, count, kind, stream)
```
kind determines the direction here, which is covered by having separate APIs in your case.  Stream isn't a request, but where the copy is performed.  From what I can see, there's no great way to determine if a specific copy has completed.  You can only query the 'stream' to determine if *all* previous operations have completed, or block on the stream until they do.

If we have a constant stream (ha) of requests flowing, it's possible that we'll never complete any of the prior ones, since the stream will never be empty.

The hsa API is more complex:
```
hsa_amd_memory_async_copy(void* dst, hsa_agent_t dst_agent, const void* src,
                              hsa_agent_t src_agent, size_t size,
                              uint32_t num_dep_signals,
                              const hsa_signal_t* dep_signals,
                              hsa_signal_t completion_signal);
```
The dep_signals parameters in this case can be used to defer the start of the copy.  I'm assuming that we don't want that and can ignore those.  The completion_signal is an event that can be updated once the copy completes.  I think this aligns with Ze events (@aingerson).

This was more aligned with what I was thinking:
User creates some event and then checks if it's been signaled.

But that flow doesn't seem to align with cuda, or even your proposal here.  Looking at the PR, it looks like you're converting the HSA model into the CUDA model.  IMO, the HSA model is better. We can check for a specific copy to have completed.  With the CUDA model, we only know if all of them have completed.  Anyway, this is what my original thoughts on an OFI abstraction for this:

```
typedef void * ofi_hmem_async_event_t;

int (*create_async_event)(device, ofi_hmem_async_event_t *event);
int (*free_async_event)(device, ofi_hmem_async_event_t event);

int (*async_copy_to_hmem)(uint64_t device, void *dest, const void *src,
	    size_t size, ofi_hmem_async_event_t event);
int (*async_copy_from_hmem)(uint64_t device, void *dest,
	const void *src, size_t size, ofi_hmem_async_event_t event);
int (*async_query_event)(ofi_hmem_async_event_t event);
```
This would allow for other async operations (if we ever need them).  It would enable checking if a single copy operation had completed, which *might* be a plus.  But also requires multiple checks to clear all async copies, which *might* be a negative.  This API should map okay to HSA and Ze, but not CUDA.  CUDA would basically convert the async_query_event() call to checking if the stream is cleared.

If this is too much overhead, we can replace async_query_event() with the more generic call like you had, async_query(), and only return if all previous async operations had completed.  If we go with the second option, I'd hide the 'stream' completely for now.  For example, we don't let the caller specify where the copy is done when calling copy_to_hmem(), so I'd align with that.  The async_query(device) would only need to know the device and would just report if everything had completed there.

Reply:
```From what I can see, there's no great way to determine if a specific copy has completed. You can only query the 'stream' to determine if all previous operations have completed, or block on the stream until they do.```

If I'm wildly off here, ignore my comment since I'm replying without context.

If you're referring to the cuda API, you can technically check if a specific copy has completed with the event API. When you call cuda memcpy async, you can record an event onto the same stream and then query/synchronize on that event to check for that specific copy. I think this maps to the query event behavior you are mentioning.

Reply:
That sounds exactly like the behavior I'm looking for.  And if OpenMPI uses this flow, then it's likely decent.  @amirshehataornl - does the above proposal seem like it would work then?

Reply:
To give a description on what exactly is done in Open MPI:

Create an array and initialize event objects on initialization (since events can be reused, don't have to call create_event/destroy_event each time). There are two pointers to keep track of first used and last used position on array.

During asynchronous memcpy, an event is selected from the array and the last pointer index is incremented % array size.  This event is then recorded on the same stream that the copy was queued on.

During progress engine phase, the first used (oldest) event is queried and if completion is detected, the first index will be incremented % array size and any copy completion operations will be performed.

Reply:
Since it may be relevant to this conversation, I spent some time writing an API in Open MPI for accelerators, similar to the hmem APIs you seem to be indicating. It may be useful to reference it and can be found here: https://github.com/open-mpi/ompi/blob/main/opal/mca/accelerator/accelerator.h

It should be pretty easy to read as all the API's are documented with doxygen format.

Reply:
> oesn't seem to align with cuda, or even your proposal here. Looking at the PR, it looks like you're converting the HSA model into the CUDA model. IMO, the HSA model is better. We can check for a specific copy to have completed. With the CUDA model, we only know if all of them ha

@shefty, Maybe there is a misunderstanding here. Unlike a cuda or hip stream, the way it's currently implemented is that each copy is queued asynchronously. The reason I group them together under the same "stream" is for the purpose of one operation which could have > 1 entries in the IOV.

For example if you try to do a async_copy_to(iovec) and the iovec.iov_len > 1, then each buffer in the IOV triggers an asynchronous copy. But the entire operation is not complete until all the buffers in the iov have been copied. 

So I'm not converting the HSA symantics into CUDA symantics.

If you have multiple calls to async_copy_to() then each one of these calls will result in a parallel operation, and will complete independently, once ROCR signals that the operation is complete.

From SHM, you'll have a set of queued operations waiting for completion. You traverse that list and query the operations, by calling the query API, and you complete the operations which are done.

So in essence, it basically works the same way as Open MPI.

What am I missing?